### ISSUE TITLE
Conditional return type in Flow?

### ISSUE THREAD

**Issue by zaynv** [facebook/flow#6055]
Is there any way to do conditional return types in Flow? I have some code (also on [flow.org/try](https://flow.org/try/#0C4TwDgpgBA8mwEsD2A7AzlAvFA3gKCijRBQGMAuKAIySQBsIBDFPAXwG488B6bqACSQB3KABMkUAJJQGwKADE6wqAGsUy4AAtGcrQgxCEdOlABOEYAFdTKKIygAFU0gC2+6AgBmUAAbEyPlD6UJ6MdGgQAPx4npZkiKhQAOYWAIIpABQ4RCSkUKyUcAnoAJS4BDIWdilYUABMAAycFV5QGf6kZfiEhOZWNtUQnISsUBDh0N09fda2Tq7uAHTmaPQAbhAZjCklw-lsXHiyUGC1KcDpm9kdlKET+SVcvFAAKprBaJpIlnSiUOpyCCmZymagQUiMSwRXwdQJCRgYCK6CR3CIAGiIEns8zc0OCMxsEFEizwYEWWggKC2NUwAD5yoQ2LsgA)): 

```ts
type Options = {
  sync: boolean
};

// How do I let Flow know that this will return a Promise if `sync` is false?
function getAge({ sync }: Options) {
  let age = 20;

  if (sync) {
    return age;
  } else {
    return Promise.resolve(age);
  }
}


let p = getAge({ sync: false })

// This should not error because `sync` was set to false, so a Promise is returned.
p.then(age => {
  
});
```

Basically, I want Flow to know that `getAge` returns a Promise if `sync` is `false`. I've tried various things, but I can't seem to get it to work.

I know I can do something like:

```ts
let p = getAge({ sync: false })

if (p instanceof Promise) {
  p.then(age => {

  });
}
```

But assuming that `getAge` is a library I was writing, I don't want to put this on the end user, and would rather annotate it myself for my library because they shouldn't have to check this themselves.

**Comment by zerobias** [comment]
```js

declare function getAge(params: {sync: true}): number;
declare function getAge(params: {sync: false}): Promise<number>;
function getAge({ sync }: Options) {
  let age = 20;

  if (sync) {
    return age;
  } else {
    return Promise.resolve(age);
  }
}```

https://flow.org/try/#0C4TwDgpgBA8mwEsD2A7AzlAvFA3gKCijRBQGMAuKAIySQBsIBDFPAXwG488ATCUuxgCdoAMwCuZRKigBzCMACCcgBRghjALZpKOYmUrBBYiKwCUlFGI1UIgnnwHCo4ychSz5SiKvVadeimdGOjQTcygABUEkDQRQgB5La1sAPjwXUil3OUUVHCISUihWSjgstFNcAigGYChGOSwoACYABk5qhBEoZQDK-EJCYWAxQXcGiE5CVigIEOgBweHR9yiYuIgAOmE0egA3bwnTKeK2LjxaqDAmnK9lfIDKEWDQ4tMuAHoPqAAVAAs4kQ-kgxHRuFAUEg6rZooJqHxGGJXgADALIqAAd0YGFCdWASCC8wANEQCYxItFYq9AcsxhBuJs8GBNsA-hAUMoJlgUlVCGxjkA

**Comment by zaynv** [comment]
Thanks a lot @zerobias! What if the `Options` type had multiple properties in it? I was able to get it to work doing something like this:

```ts
type Options = {
  name: string,
  sync: boolean
};

declare function getAge(params: Options & {sync: true }): number
declare function getAge(params: Options & {sync: false }): Promise<number>
  
function getAge({ sync }: Options) {
  let age = 20;

  if (sync) {
    return age;
  } else {
    return Promise.resolve(age);
  }
}


let p = getAge({ sync: false, name: 'hi' })

p.then(age => {
  
});
```

It seems to work, however, I'm not sure if I'm using the `&` correctly – wouldn't doing something like `Options & {sync: true }` result in 

```ts
{
  name: string,
  sync: boolean & true
}
```

which seems not right

**Comment by moroine** [comment]
Thanks for the solution. 

But it seems to have a flaw, the inner function is not properly typed. 

```javascript
type Options = {
  sync: boolean
};

declare function getAge(params: {sync: true}): number
declare function getAge(params: {sync: false}): Promise<number>
function getAge({ sync }: Options) {
  let age = 20;

  if (sync) {
    // no error: should be promise
    return Promise.resolve(age);
  } else {
    // no error: should be promise
    return age;
  }
}


let p = getAge({ sync: false })

// This should not error because `sync` was set to false, so a Promise is returned.
p.then(age => {
  
});
```
https://flow.org/try/#0C4TwDgpgBA8mwEsD2A7AzlAvFA3gKCijRBQGMAuKAIySQBsIBDFPAXwG488ATCUuxgCdoAMwCuZRKigBzCMACCcgBRghjALZpKOYmUrBBYiKwCUlFGI1UIgnnwHCo4ychSz5SiKvVadeimdGOjQTcygABUEkDQRQgB5La1sAPjwXUil3OUUVHCISUihWSjgstFNcAigGYChGOSwoACYABk5qhBEoZQDK-EJCAHohqBQkKFtowUo0AAskMTpuamgwaNjQ6sJhYDFBdyiYuIgAOmE0egA3bwaIU05CVkmQ6AHBkbGJqaQZogWlisbFB1sctoMoLt9u47o9imwuHhaiCmjkvMp8gFKCJgqFiqYuJ8ACpzOL-RbLL51H6CVakRhiPEAAwCTKgAHdGBhQnVgBMca8ADRECaMSIbE5QMlQg4QbinPBgU7AOYQFDKO5YFJVQhsB5AA
